FGEGraphics/GraphicsHelpers/Models/ShapeGenerators.cs

```diff
@@ -0,0 +1,408 @@
+            float sinTheta = (float)Math.Sin(theta);
+            float cosTheta = (float)Math.Cos(theta);
+
+            for (int j = 0; j <= sides; j++)
+            {
+                float phi = j * MathHelper.TwoPi / sides;
+                float sinPhi = (float)Math.Sin(phi);
+                float cosPhi = (float)Math.Cos(phi);
+
+                float centerX = radius * cosPhi;
+                float centerY = 0;
+                float centerZ = radius * sinPhi;
+
+                float x = (radius + tubeRadius * cosTheta) * cosPhi;
+                float y = tubeRadius * sinTheta;
+                float z = (radius + tubeRadius * cosTheta) * sinPhi;
+
+                vertices.Add(new Vector3(x, y, z));
+
+                Vector3 pointOnSurface = new(x, y, z);
+                Vector3 centerToSurface = pointOnSurface - new Vector3(centerX, centerY, centerZ);
+                normals.Add(centerToSurface.Normalized());
+
+                textureCoords.Add(new Vector2((float)j / sides, (float)i / rings));
+
+                if (i < rings && j < sides)
+                {
+                    uint currentRow = (uint)(i * (sides + 1));
+                    uint nextRow = (uint)((i + 1) * (sides + 1));
+
+                    indices[indexIndex++] = (uint)(currentRow + j);
+                    indices[indexIndex++] = (uint)(nextRow + j);
+                    indices[indexIndex++] = (uint)(nextRow + j + 1);
+
+                    indices[indexIndex++] = (uint)(currentRow + j);
+                    indices[indexIndex++] = (uint)(nextRow + j + 1);
+                    indices[indexIndex++] = (uint)(currentRow + j + 1);
+                }
+            }
+        }
+
+        if (vertices.Count != normals.Count || vertices.Count != textureCoords.Count)
+            throw new ArgumentException("Position, normal, and texture coordinate lists must have the same length.");
+
+        if (reverseOrder)
+        {
+            vertices.Reverse();
+            textureCoords.Reverse();
+            normals.Reverse();
+            Array.Reverse(indices);
+        }
+
+        return GetModelAfterGenerating(modelEngine, "torus", vertices, normals, textureCoords, indices);
+    }
+
+    private static Model GetModelAfterGenerating(
+        ModelEngine engine,
+        string name,
+        List<Vector3> vertices,
+        List<Vector3> normals,
+        List<Vector2> texCoords,
+        uint[] indices)
+    {
+        Model generatedModel = new(name)
+        {
+            Engine = engine,
+            Skinned = true,
+            ModelMin = new Location(-0.5),
+            ModelMax = new Location(0.5),
+            Original = new Model3D()
+        };
+        ModelMesh generatedModelMesh = new(name);
+        Renderable.ListBuilder builder = new();
+        builder.Prepare();
+        builder.Vertices = vertices;
+        builder.Normals = normals;
+        builder.TexCoords = texCoords.ConvertAll(t => new Vector3(t.X, t.Y, 0));
```

Comment:
This is called only by your own code, why do you need to convert anything?

Reply:
You use Vector3 for texcoords in renderable listbuilder for some reason.

Reply:
If you're building a `Model3D` anyway, just call `ModelEngine#FromScene`

Reply:
Do you mean FromSceneNoGenerate? I'm a bit confused the FromScene one looks like it generates an empty model. I took reference from GenerateCube() so that's why I did it this way.

Reply:
oooh that explains that. Yeah that code is a lil weird cause it's using the old builder api thingy - `FromScene` takes a Model3D and creates a fully working model out of it and is ideal here

Reply:
Uh it doesn't seem to be working it cant access bones on line 302 of ModelEngine so maybe we could keep it the way it is? I mean it gets the job done.